pageserver/src/tenant/storage_layer.rs

```diff
@@ -80,6 +88,103 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Default)]
+pub struct LayerAccessStats(RwLock<LayerAccessStatsInner>);
+
+#[derive(Clone)]
+struct LayerAccessStatFullDetails {
+    when: SystemTime,
+    task_kind: TaskKind,
+    access_kind: LayerAccessKind,
+}
+
+#[derive(Default)]
+struct LayerAccessStatsInner {
+    first_access: Option<LayerAccessStatFullDetails>,
+    count_by_access_kind: EnumMap<LayerAccessKind, u64>,
+    task_kind_flag: EnumSet<TaskKind>,
+    last_accesses: VecDeque<LayerAccessStatFullDetails>,
+}
+
+#[derive(Clone, Copy, strum_macros::EnumString)]
+pub enum LayerAccessStatsReset {
+    JustTaskKindFlags,
+    AllStats,
+}
+
+impl LayerAccessStatFullDetails {
+    fn to_api_model(&self) -> pageserver_api::models::LayerAccessStatFullDetails {
+        let Self {
+            when,
+            task_kind,
+            access_kind,
+        } = self;
+        pageserver_api::models::LayerAccessStatFullDetails {
+            when_millis_since_epoch: when
+                .duration_since(UNIX_EPOCH)
+                .expect("better to die in this unlikely case than report false stats")
+                .as_millis(),
+            task_kind: task_kind.into(), // into static str, powered by strum_macros
+            access_kind: *access_kind,
+        }
+    }
+}
+
+pub static LAYER_ACCESS_STATS_KILLSWITCH: AtomicBool = AtomicBool::new(false);
+
+impl LayerAccessStats {
+    fn record_access(&self, access_kind: LayerAccessKind, task_kind: TaskKind) {
+        if LAYER_ACCESS_STATS_KILLSWITCH.load(atomic::Ordering::SeqCst) {
+            return;
+        }
+        let mut inner = self.0.write().unwrap();
```

Comment:
If you're worried about lock contention, could use `try_write` here and just drop the data point if the lock is not free. The stats don't need to be 100% accurate.

Reply:
(Now a Mutex instead of RwLock.)

I think here it doesn't really matter or if it would, contention might be really bad sampling strategy. Read currently is actually more costly than this update, there are no allocations etc. cannot really see how this could create a deadlock either.

Reply:
Yeah, I'm not worried about lock contention, and dropping stats leaves a margin of uncertainty that I want to avoid.